File: JS Reverse/42 其他/加密货币/jmhb.py
"""
稍微注意一下，发现解密用到两个参数：解密数据和headers.user。后者是响应头里面的user，可通过response.headers取出，不用管是如何加密的。
"""
import requests
import execjs
import time
timer = str(int(time.time()))
headers = {
    'accept': 'application/json',
    'accept-language': 'zh-CN,zh;q=0.9,en;q=0.8',
    'cache-control': 'no-cache',
    'cache-ts': timer,
    'encryption': 'true',
    'language': 'zh',
    'origin': 'https://www.coinglass.com',
    'pragma': 'no-cache',
    'priority': 'u=1, i',
    'referer': 'https://www.coinglass.com/',
    'sec-ch-ua': '"Google Chrome";v="131", "Chromium";v="131", "Not_A Brand";v="24"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
    'sec-fetch-dest': 'empty',
    'sec-fetch-mode': 'cors',
    'sec-fetch-site': 'same-site',
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
}

def spider(page):
    global decrypted_data
    params = {
        'sort': '',
        'order': '',
        'keyword': '',
        'pageNum': str(page),
        'pageSize': '20',
        'ex': 'all',
    }

    response = requests.get('https://capi.coinglass.com/api/home/v2/coinMarkets', params=params, headers=headers)
    # print(response.text)
    # print(response.headers.get('user'))

    enc_data = response.json()
    user = response.headers.get('user')
    #
    with open('./jmhb.js', 'r', encoding='utf-8') as f:
        js_code = f.read()

    # Compile JS and prepare decrypt function
    ctx = execjs.compile(js_code)

    # Pass the response data and params to decrypt
    response_text = response.text
    try:
        decrypted_data = ctx.call('decrypt', enc_data,user)
        print("Decrypted Data:", decrypted_data)
    except Exception as e:
        print("Error during decryption:", e)

    return decrypted_data

if __name__ == '__main__':
    spider(2)

File: JS Reverse/42 其他/加密货币/jmhb.js
cryptoJS = require('../../crypto-js')
require('./env.js')
require('./loader.js')
require('./mod1.js')
// t = "ozGVG4G39PqcCyHgGGwJa6bIwF9F4vbQHoVVBaT3NNN/UDUoiTOfkPwwxCNuSzik"
// e = 'Y29pbmdsYXNzL2Fw'
eE = window.loader(27885)
eR = window.loader(81354);

ts = "https://capi.coinglass.com";
var eN = function (t) {
    var e, n = eE.ZP.inflate(new Uint8Array(t.match(/[\da-f]{2}/gi).map(function (t) {
        return parseInt(t, 16)
    }))), r = "", i = 16384;
    for (e = 0; e < n.length / i; e++)
        r += String.fromCharCode.apply(null, n.slice(e * i, (e + 1) * i));
    return decodeURIComponent(escape(r += String.fromCharCode.apply(null, n.slice(e * i))))
}
eB = function (t) {
    var e = t;
    return -1 != t.indexOf("ipa/".split("").reverse().join("")) && (e = -1 == t.indexOf("?") ? t.substring(t.indexOf("ipa/".split("").reverse().join("")), t.length) : t.substring(t.indexOf("ipa/".split("").reverse().join("")), t.indexOf("?"))),
        e
}

eZ = function (t, e) {
    var n = eN(eR.AES.decrypt(t, eR.enc.Utf8.parse(e), {
        mode: eR.mode.ECB,
        padding: eR.pad.Pkcs7
    }).toString(eR.enc.Hex));
    return '"' == n.charAt(0) && (n = n.substring(1, n.length)),
    '"' == n.charAt(n.length - 1) && (n = n.substring(0, n.length - 1)),
        n
}

// console.log(eZ(t, e))


function decrypt(data,user) {
    // var e = eB(t.config.url)
    let t = {
        "data": data,
        "status": 200,
        "statusText": "",
        "headers": {
            "cache-control": "no-cache, no-store, must-revalidate",
            "content-type": "application/json",
            "encryption": "true",
            "user": user
        },
        "config": {
            "transitional": {
                "silentJSONParsing": true,
                "forcedJSONParsing": true,
                "clarifyTimeoutError": false
            },
            "adapter": [
                "xhr",
                "http"
            ],
            "transformRequest": [
                null
            ],
            "transformResponse": [
                null
            ],
            "timeout": 10000,
            "xsrfCookieName": "XSRF-TOKEN",
            "xsrfHeaderName": "X-XSRF-TOKEN",
            "maxContentLength": -1,
            "maxBodyLength": -1,
            "env": {},
            "headers": {
                "Accept": "application/json",
                "language": "zh",
                "encryption": "true",
                "cache-ts": "1732030760401"
            },
            "params": {
                "sort": "",
                "order": "",
                "keyword": "",
                "pageNum": 3,
                "pageSize": 20,
                "ex": "all"
            },
            "method": "get",
            "url": "https://capi.coinglass.com/api/home/v2/coinMarkets",
            "baseURL": "https://capi.coinglass.com"
        },
        "request": {},
        "timeout": 30000
    }
    let e = "/api/home/v2/coinMarkets"
        , n = btoa("coinglass".concat(e, "coinglass"));
    console.log('n:::', n)
    let c = n.substring(0, 16),
        s = eZ(t.headers.user, c);
    console.log('c:::', c)
    console.log('s:::', s)
    return eZ(t.data.data, s)
}

// 接口测试


// console.log(decrypt(t))


